packages/ens-controller/src/EnsController.ts

```diff
@@ -86,6 +111,32 @@ export class EnsController extends BaseControllerV2<
         ...state,
       },
     });
+
+    if (provider && onNetworkDidChange && getCurrentChainId) {
```

Comment:
There are a couple of things I'm noticing about this section.

1. This controller gets the current chain ID, builds an instance of Ens, then does the same thing when the network changes. This seems a bit repetitious to me. I believe that we have to do things this way because we're using `onNetworkDidChange`, which we can assume gets called when the `networkDidChange` event occurs. However, this event doesn't help us that much because it does not occur when the network is initialized, but only after the network is switched following initialization.

2. Using the chain ID to get the network ID seems error-prone to me. These two concepts aren't the same and we shouldn't be assuming that there is a mapping from one to the other (since that's not in a standard anywhere). We are also using a function to get the network ID which is an old way of doing things.

To solve both of these problems, I think we ought to use `onNetworkStateChange` instead of `onNetworkDidChange`. We could hook this up when EnsController is initialized to get called every time the network state is changed, which happens even after the NetworkController is initialized. This would allow us to write the code that sets `this._ens` only once.

Also, since the NetworkController in this repo relies on BaseControllerV2, I think it'd be better to add a controller action to NetworkController which would allow us to get the current network ID. You can see an example of a controller action in NetworkController here:
 https://github.com/MetaMask/core/blob/cb43d9e581e2b3d0be09c2fe54d81cc4c9f4b146/packages/network-controller/src/NetworkController.ts#L234

Here we are registering two actions. When a consumer says `messenger.call('NetworkController:getProviderConfig')`, they will get the current provider config object; and when they say `messenger.call('NetworkController:getEthQuery')` they will get the EthQuery instance that points to the currently selected network.

   Following this example, what if we updated NetworkController in this repo to register a third action handler called, say, `NetworkController:getNetworkId`? It would be a bit tricky here because the way we store network ID in NetworkController is strange, but we could have the action handler return something like `this.state.network === 'loading' ? null : this.state.network`. And in the extension we could do a similar thing.

Once we have those pieces in place, the code here could look something like:

``` typescript
      this._ens = null;

      let previousNetworkId = null;
      onNetworkStateChange((state) => {
        const currentNetworkId = messenger.call('NetworkController:getNetworkId');
        if (currentNetworkId !== previousNetworkId) {
          if (Ens.getNetworkEnsSupport(currentNetworkId)) {
            this._ens = new Ens({ network: networkId, provider });
          } else {
            this._ens = null;
          }
          previousNetworkId = currentNetworkId;
        });
      });
```

Reply:
Oh I see, the map we're using is keyed by network ID. This is a relic from before we switched over to chain ID years ago; we can convert the mapping to be keyed by chain ID as well. That would be more accurate; the ENS contract doesn't exist across all networks with that ID, it only exists on the specific listed chains.

Reply:
Okay. I was really curious whether we could get rid of this too, but didn't have the proper context behind these ENS changes so I wasn't sure how to advise on this. I can take a closer look at this.